common/logger/src/lib.rs

```diff
@@ -3,131 +3,361 @@
+    fn log(&self, record: &Record) {
+        // The following filters out everything that does not deal with Libra
+        if record.metadata().target() != DEFAULT_TARGET {
+            return;
+        }
+
+        if !self.filter.matches(record) {
+            return;
+        }
+
+        match format(record) {
+            Ok(formatted) => self.writer.write(formatted),
+            Err(e) => self
+                .writer
+                .write(format!("Unable to format log {:?}: {}", record, e)),
+        };
+    }
+
+    /// In this logger, this doesn't do anything.
+    fn flush(&self) {}
+}
+
+/// Operations between the AsyncLogClient and AsyncLogService
+enum LogOp {
+    /// Send a line to log to the writer.
+    Log(String),
+}
+
+/// Provides the backend to the asynchronous interface for logging. This part actually writes the
+/// log to the writer.
+struct AsyncLogService<W> {
+    receiver: Receiver<LogOp>,
+    writer: W,
+}
+
+impl<W: Writer> AsyncLogService<W> {
+    /// Loop for handling logging
+    fn log_handler(mut self) {
+        loop {
+            if let Err(e) = self.handle_recv() {
+                // TODO write an error record and then exit
+                self.writer.write(format!("Unrecoverable error: {}", e));
+                return;
+            }
+        }
+    }
+
+    fn handle_recv(&mut self) -> Result<(), RecvError> {
+        match self.receiver.recv()? {
+            LogOp::Log(line) => self.writer.write(line),
+        };
+        Ok(())
+    }
+}
+
+/// Provides the log::Log interface for Libra's asynchronous logger
+struct AsyncLogClient {
+    filter: filter::Filter,
+    sender: SyncSender<LogOp>,
+}
+
+impl Log for AsyncLogClient {
+    /// Determines if a log message with the specified metadata would be logged.
+    fn enabled(&self, metadata: &Metadata) -> bool {
+        self.filter.enabled(metadata)
+    }
+
+    /// Logs the provided record but first evaluates the filters and then sending it to the
+    /// AsyncLogService via a SyncSender.
+    fn log(&self, record: &Record) {
+        // The following filters out everything that does not deal with Libra
+        if record.metadata().target() != DEFAULT_TARGET {
+            return;
+        }
+
+        if !self.filter.matches(record) {
+            return;
+        }
+
+        let formatted = match format(record) {
+            Ok(formatted) => formatted,
+            Err(e) => format!("Unable to format log {:?} due to {}", record, e),
         };
-        Logger::root(async_builder.build().fuse(), o!())
-    } else {
-        Logger::root(Mutex::new(drain).fuse(), o!())
+        if let Err(e) = self.sender.try_send(LogOp::Log(formatted)) {
+            match e {
+                TrySendError::Disconnected(_) => {
+                    eprintln!("Unable to log {:?} due to {}", record, e)
+                }
+                TrySendError::Full(_) => eprintln!("Unable to log, queue full"),
```

Comment:
This probably needs some better handling, for example having atomic counter in `LogService` and incrementing counter here, and then printing message like `Skipped {counter} log entries`
Current implementation basically goes from async logger to sync logger when queue is full. It means that instead of quickly dropping logs, all threads will start to compete for lock on stderr.

Reply:
Yes, let me massage that a little bit in a follow up.